File: libdriveless/tests/src/test_search_frame.cpp
#include <gtest/gtest.h>
#include "../../include/search_frame.h"
#include "test_utils.h"
#include <cmath>

TEST(TestSearchFrame, TestSetGet)
{
    SearchFrame f1(1000, 1001, {50, 51}, {150, 151});

    ASSERT_EQ(1000, f1.width());
    ASSERT_EQ(1001, f1.height());

    // ASSERT_EQ(5, f1.minDistance().first);
    // ASSERT_EQ(6, f1.minDistance().second);

    ASSERT_EQ(50, f1.lowerBound().first);
    ASSERT_EQ(51, f1.lowerBound().second);

    ASSERT_EQ(150, f1.upperBound().first);
    ASSERT_EQ(151, f1.upperBound().second);
}

TEST(TestSearchFrame, TestConvertColors)
{
    SearchFrame f1(100, 100, {5, 5}, {15, 15});

    int c = 0;
    for (int i = 0; i < 100; i++)
        for (int j = 0; j < 100; j++)
        {
            f1[{j, i}].x = c;
            c = (c + 1) % 29;
        }

    std::vector<std::tuple<int, int, int>> colors = {
        {0, 0, 0},
        {128, 64, 128},
        {244, 35, 232},
        {70, 70, 70},
        {102, 102, 156},
        {190, 153, 153},
        {153, 153, 153},
        {250, 170, 30},
        {220, 220, 0},
        {107, 142, 35},
        {152, 251, 152},
        {70, 130, 180},
        {220, 20, 60},
        {255, 0, 0},
        {0, 0, 142}, // car
        {0, 0, 70},
        {0, 60, 100},
        {0, 80, 100},
        {0, 0, 230},
        {119, 11, 32},
        {110, 190, 160},
        {170, 120, 50},
        {55, 90, 80},
        {45, 60, 150},
        {157, 234, 50},
        {81, 0, 81},
        {150, 100, 100},
        {230, 150, 140},
        {180, 165, 180}};

    f1.setClassColors(colors);

    uchar *dest = new uchar[3 * 100 * 100];
    f1.exportToColorFrame(dest);

    c = 0;
    int x, y, z;
    for (int i = 0; i < 100; i++)
        for (int j = 0; j < 100; j++)
        {
            std::tie(x, y, z) = colors[c];
            int pos = 3 * (i * 100 + j);

            if (dest[pos] != x || dest[pos + 1] != y || dest[pos + 2] != z)
            {
                printf("error in position (%d, %d): expected (%d, %d, %d), obtained (%d, %d, %d)\n", i, j, x, y, z, dest[pos], dest[pos + 1], dest[pos + 2]);
                FAIL();
            }

            c = (c + 1) % 29;
        }
}

TEST(TestSearchFrame, TestClassCosts)
{
    SearchFrame f1(100, 100, {5, 5}, {15, 15});

    std::vector<float> costs({{0.0},
                              {1.0},
                              {2.0},
                              {3.1},
                              {4.2},
                              {5.3}});
    f1.setClassCosts(costs);

    ASSERT_DEQ(f1.getClassCost(0), 0.0);
    ASSERT_DEQ(f1.getClassCost(1), 1.0);
    ASSERT_DEQ(f1.getClassCost(2), 2.0);
    ASSERT_DEQ(f1.getClassCost(3), 3.1);
    ASSERT_DEQ(f1.getClassCost(4), 4.2);
    ASSERT_DEQ(f1.getClassCost(5), 5.3);
}

File: libdriveless/tests/python/test_search_frame.py
import sys, time
sys.path.append("../../")
sys.path.append("../")
import unittest, math, numpy as np
import matplotlib.pyplot as plt
from pydriveless import SearchFrame, float3

import cv2
from test_utils import fix_cv2_import
fix_cv2_import()


COLORS = np.array([
            [0, 0, 0],
            [128, 64, 128],
            [244, 35, 232],
            [70, 70, 70],
            [102, 102, 156],
            [190, 153, 153],
            [153, 153, 153],
            [250, 170, 30],
            [220, 220, 0],
            [107, 142, 35],
            [152, 251, 152],
            [70, 130, 180],
            [220, 20, 60],
            [255, 0, 0],
            [0, 0, 142], # car
            [0, 0, 70],
            [0, 60, 100],
            [0, 80, 100],
            [0, 0, 230],
            [119, 11, 32],
            [110, 190, 160],
            [170, 120, 50],
            [55, 90, 80],
            [45, 60, 150],
            [157, 234, 50],
            [81, 0, 81],
            [150, 100, 100],
            [230, 150, 140],
            [180, 165, 180]
        ], dtype=np.uint32)

class TestSearchFrame(unittest.TestCase):
        
    def test_get_set(self):
        frame = SearchFrame(100, 101, (45, 55), (55, 45))
        self.assertEqual(frame.width(), 100)
        self.assertEqual(frame.height(), 101)
        self.assertEqual(frame.lowerBound()[0], 45)
        self.assertEqual(frame.lowerBound()[1], 55)
        self.assertEqual(frame.upperBound()[0], 55)
        self.assertEqual(frame.upperBound()[1], 45)
        
    
    def test_export_color_frame(self):
        frame = SearchFrame(100, 101,  (45, 55), (55, 45))
        frame.set_class_colors(COLORS)       
        data = np.zeros((100, 100, 3), dtype=np.float32)
        c = 0
        for i in range(100):
            for j in range(100):
                data[i, j] = c
                c = (c + 1) % 29
                
            
        frame.set_frame_data(data)
        f = frame.get_color_frame()
        
        for i in range(100):
            for j in range(100):
                for k in range(3):
                    f[i, j, k] == COLORS[c][k]
                c = (c + 1) % 29

    # def test_export_cuda_frame(self):
    #     frame = SearchFrame(100, 101,  (45, 55), (55, 45))
    #     frame.set_class_colors(COLORS)       
    #     data = np.zeros((100, 100, 3), dtype=np.float32)
    #     c = 0
    #     for i in range(100):
    #         for j in range(100):
    #             data[i, j] = c
    #             c = (c + 1) % 29
                
            
    #     frame.set_frame_data(data)
    #     f = frame.get_frame()
        
    #     for i in range(100):
    #         for j in range(100):
    #             f[i, j, 0] == (c + 1) % 29
                
    
    def test_convert_bev_file(self):
        fin = np.array(cv2.imread("bev_1.png"))
        frame = SearchFrame(fin.shape[1], fin.shape[0], (0, 0), (0, 0))
        frame.set_frame_data(fin)
        frame.set_class_colors(COLORS)
        fout = frame.get_color_frame()
        cv2.imwrite("bev_1_conv.png", fout)
    
    def test_class_costs(self):
        frame = SearchFrame(100, 101, (45, 55), (55, 45))
        costs = np.array([
            1.1, 2.2, 3.3, 4.4, 5.5
        ], np.float32)
        frame.set_class_costs(costs)
        
        self.assertAlmostEqual(frame.get_class_cost(0), 1.1, places=3)
        self.assertAlmostEqual(frame.get_class_cost(1), 2.2, places=3)
        self.assertAlmostEqual(frame.get_class_cost(2), 3.3, places=3)
        self.assertAlmostEqual(frame.get_class_cost(3), 4.4, places=3)
        self.assertAlmostEqual(frame.get_class_cost(4), 5.5, places=3)

    def test_get_set(self):
        frame = SearchFrame(100, 101, (-1, -1), (-1, -1))
        data = np.full((100, 100, 3), 1.0, np.float32)
        frame.set_frame_data(data)
        
        frame[(30, 30)] = float3(2.0, 1.0, 1.0)
        
        for z in range(100):
            for x in range(100):
                if x == 30 and z == 30:
                    self.assertAlmostEqual(2.0, frame[(x, z)].x)
                elif  frame[(x, z)].x != 1.0 or\
                    frame[(x, z)].y != 1.0 or\
                    frame[(x, z)].z != 1.0:
                        self.fail(f"pos {x}, {x} is wrong: should be (1.0, 1.0, 1.0) but it is {frame[(x, z)]}")

    

if __name__ == "__main__":
    unittest.main()
